src/main/java/com/machowina/service/TicketServiceImp.java

```diff
@@ -0,0 +1,181 @@
+		
+		ParkingTicket newTicket = new ParkingTicket(startTime, zone, car, driver);
+		
+		return newTicket;
+		
+	}
+
+	@Override
+	public void checkForDuplicatingTicket(Long carId) {
+		ParkingTicket ticket = ticketRepository.findOneByCarIdAndIsStoppedFalse(carId);
+		if (ticket != null) {
+			throw new TicketAlreadyRunning();
+		}
+	}
+	
+
+	@Override
+	@Transactional
+	public Long saveTicket(ParkingTicket ticket) {
+		ParkingTicket savedTicket = ticketRepository.save(ticket);
+		return savedTicket.getId();
+	}
+
+	@Override
+	@Transactional
+	public void stopTicket(Long ticketId) {
+		ParkingTicket ticket = findById(ticketId);
+		if (!ticket.isStopped()) {
+			ticket.setStopTime(LocalDateTime.now());
+			ticket.setStopped(true);
+			saveTicket(ticket);
+		} else {
+			throw new TicketAlreadyStopped();
+		}
+	
+	
+	}
+
+	@Override
+	public ParkingTicket findById(Long ticketId) {
+		ParkingTicket ticket = ticketRepository.findOne(ticketId);
+		if (ticket == null) {
+			throw new EntityNotFoundException("There is no ticket with this ID");
+		} else {
+			return ticket;
+		}
+	}
+
+	
+	
+
+	@Override
+	public boolean checkForValidTicket(String carLicense, Long parkingZoneId) {
+		List <ParkingTicket> validTickets = ticketRepository
+				.findByCarLicenseAndParkingZoneIdAndIsStoppedFalse(carLicense, parkingZoneId);
+		
+		return (!validTickets.isEmpty());
+	}
+
+	
+
+	@Override
+	@Transactional
+	public BigDecimal calculateTicketFee(Long ticketId) {
+		ParkingTicket ticket = findById(ticketId);
+		if ((!ticket.isStopped())||(ticket.getStopTime()==null)) {
+			throw new TicketNotStoppedException();
+		}
+		ParkingRates rates = parkingRatesService.findOneByDriverTypeAndParkingZone
+				(ticket.getDriver().getDriverType(), ticket.getParkingZone());
+		
+		Duration parkingTime = Duration.between(ticket.getStartTime(), ticket.getStopTime());
+		BigDecimal fee = calculateFeeForGivenDuration(parkingTime, rates);
+		
+		ticket.setFee(fee);
+		saveTicket(ticket);
+		return fee;
+	}
+
+	@Override
+	public BigDecimal calculateFeeForGivenDuration(Duration parkingTime, ParkingRates rates) {
+		
+		Long parkingHours = parkingTime.toHours();
+		Long incompleteHourMinutes = parkingTime.toMinutes() - parkingHours*60;
+		BigDecimal fee = BigDecimal.ZERO;
+		BigDecimal thisHourFee = BigDecimal.ZERO;
+		BigDecimal incompleteHourRate = BigDecimal.ZERO;
+		
+		for (int i = 0; i <= parkingHours; i++) {
+			switch (i) {
+			case 0:
+				incompleteHourRate = rates.getFirstHourCost();
+				break;
+			case 1:
+				thisHourFee = rates.getFirstHourCost();
+				incompleteHourRate = rates.getSecondHourCost();
+				break;
+			case 2:
+				thisHourFee = rates.getSecondHourCost();
+				incompleteHourRate = thisHourFee.multiply(rates.getNextHourRate());
+				break;
+			default:
+				thisHourFee = thisHourFee.multiply(rates.getNextHourRate());
+				incompleteHourRate = thisHourFee.multiply(rates.getNextHourRate());
+				break;
+			}
+			fee = fee.add(thisHourFee);
+		}
+		
```

Comment:
czy to jest dobre miejsce na taką logikę?

Reply:
Raczej nie, możnaby to umieścić w klasie ParkingRates